### PR TITLE
Updated Tenant Controller to Match with the Sample-Controller

### PR DESCRIPTION
**Pull request by ubombar**

### CODE REVIEW

**Comment by ubombar** [comment]
Hello, I am still working on the tenant controller since there are some issues on the other files such as userrequest and tenantrequest. Also I have a minor bug when I run the test case. I will be adding more commits as I resolve the issues.

File: cmd/tenant/main.go [hunk 1]
```diff
@@ -2,12 +2,19 @@ package main
...
 	"github.com/EdgeNet-project/edgenet/pkg/controller/core/v1alpha/tenant"
+	informers "github.com/EdgeNet-project/edgenet/pkg/generated/informers/externalversions"
+	"github.com/EdgeNet-project/edgenet/pkg/signals"
+
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/klog"
 )
 
 func main() {
+	stopCh := signals.SetupSignalHandler()
```
Comment (bsenel) [inline cmd/tenant/main.go:22]: You should add `klog.InitFlags(nil)` and `flag.Parse()` according to how to use klog description.

Comment (bsenel) [reply cmd/tenant/main.go:22]: Would you please put them above `stopCh := signals.SetupSignalHandler()`?

File: pkg/controller/apps/v1/nodelabeler/controller.go [hunk 2]
```diff
@@ -1,204 +1,165 @@
...
-	var err error
-
-	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
-		utilruntime.HandleError(err)
-		return
-	}
-
-	c.workqueue.Add(key)
+	return true
 }
```
Comment (bsenel) [inline pkg/controller/apps/v1/nodelabeler/controller.go]: It seems like the old version of the node labeler controller.

File: pkg/controller/apps/v1/nodelabeler/controller_test.go [hunk 3]
```diff
@@ -2,115 +2,34 @@ package nodelabeler
...
-
-		newController := NewController(kubeclientset,
-			kubeInformerFactory.Core().V1().Nodes())
-
-		kubeInformerFactory.Start(stopCh)
-		controller = newController
-		if err := controller.Run(2, stopCh); err != nil {
-			klog.Fatalf("Error running controller: %s", err.Error())
-		}
-	}()
```
Comment (bsenel) [inline pkg/controller/apps/v1/nodelabeler/controller_test.go]: Previous version?

File: pkg/controller/apps/v1/nodelabeler/handler.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		log.Infof("External IP: %s", externalIP)
+		result = node.GetGeolocationByIP(obj.(*corev1.Node).Name, externalIP)
+	}
+	// Check if the internal IP exists and
+	// the result of detecting geolocation by external IP is false
+	if internalIP != "" && result == false {
+		log.Infof("Internal IP: %s", internalIP)
+		node.GetGeolocationByIP(obj.(*corev1.Node).Name, internalIP)
+	}
+}
```
Comment (bsenel) [inline pkg/controller/apps/v1/nodelabeler/handler.go]: Yes, not it is clear that the pull request includes the previous version of the node labeler controller.

File: pkg/controller/apps/v1/nodelabeler/nodelabeler_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+// The main structure of test group
+type testGroup struct {
+	client  kubernetes.Interface
+	handler Handler
+	nodeObj corev1.Node
+}
+
+func TestMain(m *testing.M) {
+	flag.String("geolite-path", "../../../../../assets/database/GeoLite2-City/GeoLite2-City.mmdb", "Set GeoIP DB path.")
```
Comment (bsenel) [inline pkg/controller/apps/v1/nodelabeler/nodelabeler_test.go]: Please look at all the files related to node labeler.

File: pkg/controller/registration/v1alpha/tenantrequest/handler.go [hunk 6]
```diff
@@ -95,7 +95,7 @@ func (t *Handler) ObjectCreatedOrUpdated(obj interface{}) {
 								user.Spec.LastName = tenantRequest.Spec.Contact.LastName
 								user.Spec.Role = "Owner"
 								user.SetLabels(map[string]string{"edge-net.io/user-template-hash": util.GenerateRandomString(6)})
-								tenantHandler.ConfigurePermissions(tenant, user.DeepCopy(), tenantv1alpha.SetAsOwnerReference(tenant))
+								permission.ConfigureTenantPermissions(tenant, user.DeepCopy(), tenantv1alpha.SetAsOwnerReference(tenant), t.edgenetClientset)
```
Comment (bsenel) [inline pkg/controller/registration/v1alpha/tenantrequest/handler.go]: You might consider putting `var EdgenetClientset clientset.Interface` as a global variable into the permission package and sync that in the `NewController` method.

File: pkg/controller/registration/v1alpha/userrequest/handler.go [hunk 7]
```diff
@@ -80,9 +80,7 @@ func (t *Handler) ObjectCreatedOrUpdated(obj interface{}) {
 		if tenant.Spec.Enabled {
 			if userRequest.Spec.Approved {
 				userRequest.SetLabels(map[string]string{"edge-net.io/user-template-hash": util.GenerateRandomString(6)})
-				tenantHandler := tenantv1alpha.Handler{}
-				tenantHandler.Init(t.clientset, t.edgenetClientset)
-				tenantHandler.ConfigurePermissions(tenant, userRequest, tenantv1alpha.SetAsOwnerReference(tenant))
+				permission.ConfigureTenantPermissions(tenant, userRequest, tenantv1alpha.SetAsOwnerReference(tenant), t.edgenetClientset)
```
Comment (bsenel) [inline pkg/controller/registration/v1alpha/userrequest/handler.go]: Same as above.

File: pkg/permission/permission.go [hunk 8]
```diff
@@ -20,15 +20,39 @@ import (
...
+	"aup-rolebinding-failure":           "AUP role binding creation failed, user: %s",
+	"permission-rolebinding-failure":    "Permission role binding creation failed, user: %s",
+	"administrator-rolebinding-failure": "Administrator role binding creation failed, user: %s",
+	"user-failure":                      "User creation failed due to lack of labels, user: %s",
+	"cert-failure":                      "Client cert generation failed, user: %s",
+	"kubeconfig-failure":                "Kubeconfig file creation failed, user: %s",
+}
+
 // Clientset to be synced by the custom resources
 var Clientset kubernetes.Interface
```
Comment (bsenel) [inline pkg/permission/permission.go:57]: You can define the global variable for the EdgeNet client here.

File: pkg/registration/registration.go [hunk 9]
```diff
@@ -298,3 +301,29 @@ func CreateConfig(serviceAccount *corev1.ServiceAccount) string {
 	}
 	return string(dat)
 }
+
+// Create function is for being used by other resources to create a tenant
+func CreateTenant(obj interface{}, edgenetclientset clientset.Interface) bool {
```
Comment (bsenel) [inline pkg/registration/registration.go]: Please define the global variable for `edgenetclientset` at line 59.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 10]
```diff
@@ -17,48 +17,67 @@ limitations under the License.
...
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
 }
 
 // Constant variables for events
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: I prefer moving this part, until line 94, to above the Controller struct type.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 11]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
-			// Put the resource object into a key
-			event.key, err = cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-			event.function = delete
-			log.Infof("Delete tenant: %s", event.key)
-			if err == nil {
-				queue.Add(event)
+				controller.handleObject(newObj)
 			}
 		},
+		DeleteFunc: controller.handleObject,
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: Please use `enqueueTenant` to replace `handleObject` here.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 12]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
+
+// Run will set up the event handlers for the types of selective deployment and node, as well
+// as syncing informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	klog.V(4).Infoln("Starting Selective Deployment controller")
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: I see that the Selective Deployment term exists in so many places.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 13]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
 
 // This function deals with the queue and sends each item in it to the specified handler to be processed.
-func (c *controller) processNextItem() bool {
-	log.Info("processNextItem: start")
-	// Fetch the next item of the queue
-	event, quit := c.queue.Get()
-	if quit {
+func (c *Controller) processNextWorkItem() bool {
+	// Fetch the item from workqueue
+	event, shutdown := c.workqueue.Get()
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: Please replace `event, shutdown` with `obj, shutdown`.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 14]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
-			c.queue.AddRateLimited(event.(informerevent).key)
-		} else {
-			c.logger.Errorf("Controller.processNextItem: Failed processing item with key %s with error %v, no more retries", event.(informerevent).key, err)
-			c.queue.Forget(event.(informerevent).key)
-			utilruntime.HandleError(err)
+		// In this case we assume the resource is deleted
+		if errors.IsNotFound(err) {
+			return nil
 		}
+		utilruntime.HandleError(fmt.Errorf("tenant '%s' in work queue no longer exists", name))
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: It should be above line 273.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 15]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
+		return nil
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: return err.

File: pkg/controller/core/v1alpha/tenant/controller.go [hunk 16]
```diff
@@ -74,151 +93,327 @@ var statusDict = map[string]string{
...
 	}
 
-	if !exists {
-		if event.(informerevent).function == delete {
-			c.logger.Infof("Controller.processNextItem: object deleted detected: %s", keyRaw)
-			c.handler.ObjectDeleted(item)
+	c.ObjectCreatedOrUpdated(tenant)
```
Comment (bsenel) [inline pkg/controller/core/v1alpha/tenant/controller.go]: Let's find a proper name for this method. `TuneTenant()`?

**Review by bsenel — CHANGES_REQUESTED** [review]
Good start! 

I found the Selective Deployment term in many places. Would you mind replacing them with Tenant? In addition to this, it would be great to follow the same order for the methods in the controllers. 

I also put some comments for you to look at.

**Review by bsenel — APPROVED** [review]
It seems OK now, so it goes into release-1.0.

**Pull request merged**